### ISSUE TITLE
Angular 8/lazy modules/unsafe scopes are forbidden

### ISSUE THREAD

**Issue by dmytro-gokun** [cartant/tslint-etc#23]
Angular 8.0.0 changed the way lazy modules are loaded. Now, the code like this:

```
loadChildren: () => import('../login/login.module').then(m => m.LoginModule)
```

Produces warning "Unsafe scopes are forbidden" (with "no-unsafe-callback-scope": true). I'm not sure what is the potential problem here... Can someone explain?

**Comment by cartant** [comment]
If the line highlighted by the lint failure is the line you included in your bug report, then this is a bug. However, that snippet alone is insufficient to effect the failure. Please include an entire, minimal file that effects the linting failure.

**Comment by cartant** [comment]
Note that this code does not effect the failure:

```ts
import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

const routes: Routes = [
  {
    path: 'lazy',
    loadChildren: () => import('./lazy/lazy.module').then(m => m.LazyModule)
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule {}
```

**Comment by dmytro-gokun** [comment]
@cartant Yep, i see. Okay, I will try to build a minimal replication and post it here a bit later

**Comment by dmytro-gokun** [comment]
@cartant Okay, I've built a reproduction. Please find the attached project. After unzipping, run "npm ci", then "npm run lint". Let me know if this reproduces the issue.

[tslint-etc-issue23.zip](https://github.com/cartant/tslint-etc/files/3289651/tslint-etc-issue23.zip)

NB. I've created this by running "ng new" using the latest Angular/CLI (8.0.3) and then adding tsling-etc references & rules + adding a couple of configuration settings to tsconfig. So, it has nothing more then a usual fresh Angular project would have.

**Comment by cartant** [comment]
Awesome. Thanks for taking the time to put it together. I'll have a look at it on the weekend.

**Comment by cartant** [comment]
FYI, I've been able to reproduce the problem.  To effect the error, in the fixture, it's necessary to have a file to which the dynamic import resolves. Thanks.

**Comment by dmytro-gokun** [comment]
Great. Thanks for the update. I hope it is not super-complicated to fix.

**Comment by cartant** [comment]
This should be fixed in 1.5.4. It was much harder to fix than I'd anticipated. I ended up re-writing the `unsafe-scope` rule in the RxJS rules and the re-implementing this one based on the new implementation.

**Comment by dmytro-gokun** [comment]
@cartant Okay, i think the rewrite has introduced a new issue. Now I'm getting the same warning when accessing global objects:

```
platformBrowserDynamic()
    .bootstrapModule(AppModule)
    .catch(err => console.log(`ERROR CAUGHT: ${err}`)); <-- HERE
```

```
    @Effect({ dispatch: false })
    onSetToken$ = this.actions$.pipe(
        ofType<SetToken>(SetToken.TYPE),
        tap(action => {
            localStorage.setItem(TOKEN_NAME, action.token);   <-- HERE
            if (action.url !== undefined) {
                this.router.navigateByUrl(action.url);
            }
        })
    );
```

I do not this this lines produced this warning before the last fix. Can you plz take a look at this?

**Comment by cartant** [comment]
That's cause `console` is not in this list, yet: https://github.com/cartant/tslint-etc/blob/49c4b57bd46ca95082f90fb6aa94db3d0242e36e/source/rules/noUnsafeCallbackScopeRule.ts#L20

And if you are getting that error in `tap`, you should use `"allowTap": true` in the options - which is the default.

**Comment by cartant** [comment]
TBH, I am giving serious consideration to deleting said rules. ATM, I think they are best used as warnings, if at all.

**Comment by dmytro-gokun** [comment]
@cartant 

> And if you are getting that error in tap, you should use "allowTap": true in the options - which is the default.

Well, i have this in my tslint.congif: "rxjs-no-unsafe-scope": true, so  "allowTap" should be set to true. And, in fact, it does not compain about "tap". It complains for the line which has access to "localStorage".

> TBH, I am giving serious consideration to deleting said rules. ATM, I think they are best used as warnings, if at all.

Yeah, i'm not sure if this rule is helpful as well. I've never caught any problem in the code using it, as far as remember. However, may be that's because my coding style does not produce much fuel for the rule. Maybe, in the general case, it is still useful.

**Comment by cartant** [comment]
You really should not use both rules. They are checking the same thing, but the rule in `tslnt-etc` is more general, so any settings in the RxJS are largely irrelevant.

TBH, you are better off creating a new issue, too, as now I don't know which to which rule you are referring. Anyway, I won't have time to look at this for at least a few days.

**Comment by dmytro-gokun** [comment]
@cartant Okay, my bad. I'm not sure why i mentioned "rxjs-no-unsafe-scope" at all, i meant to mention "no-unsafe-callback-scope". If you'd like it's not a problem for me to create a new issue. even though see this new warning as the result of latest fix you did (they were not present before, as far as I remember).

**Comment by cartant** [comment]
> ... even though see this new warning as the result of latest fix you did.

Doesn't matter. As a general rule, it's always a good idea to create a new issue rather than comment on one that's been closed. Especially as:

> ... this new warning ...

Just reference the old issue from the new one. Dealing with comments on closed issues is a PITA for maintainers.
